[PowerPC] Add an alias for -mregnames so that full register names used in assembly.

stefanp-synopsys opened the pull request:
This option already exists on GCC and so it is being added to LLVM so that we use the same option as them.

stefanp-synopsys committed:
[PowerPC] Add an alias for -mregnames so that full register names used in assembly.

This option already exists on GCC and so it is being added to LLVM so that we
use the same option as them.

stefanp-synopsys committed:
Missed a little cleanup on a couple of source files.

chenzheng1030 commented on clang/lib/Basic/Targets/PPC.h:
```diff
@@ -80,6 +80,7 @@ class LLVM_LIBRARY_VISIBILITY PPCTargetInfo : public TargetInfo {
   bool IsISA3_0 = false;
   bool IsISA3_1 = false;
   bool HasQuadwordAtomics = false;
+  bool FullRegisterNames = false;
```
Using a target feature bit for assembly printing seems not match other bits. IIUC, all the bits here should control the available instructions on a subtarget.

Could we use an option in `TargetMachine::Options::MCOptions`? this looks like the way where clang can accept an option and control the code generation in the backend. The current option `Options.MCOptions.PreserveAsmComments` seems a little similar with this functionality.

stefanp-ibm replied:
I did take a look at trying to do it this way at first but I was concerned that the option I wanted to add was target specific and I didn't want to add an option to `MCAsmInfo.h` that was only for PPC. However, I've taken a look at the existing options in there and it seems that there are already a set of options that are specific to one target or another so I'll try this approach.

MaskRay commented on clang/include/clang/Driver/Options.td:
```diff
@@ -4798,6 +4798,10 @@ def mrop_protect : Flag<["-"], "mrop-protect">,
 def mprivileged : Flag<["-"], "mprivileged">,
     Group<m_ppc_Features_Group>;
+def mregnames : Flag<["-"], "mregnames">, Group<m_ppc_Features_Group>,
+                Visibility<[ClangOption]>;
```
`Visibility<[ClangOption]>` is the default and should be omitted.

MaskRay commented on clang/test/CodeGen/PowerPC/ppc-full-reg-names.c:
```diff
@@ -0,0 +1,78 @@
+// REQUIRES: powerpc-registered-target
```
Tests at the wrong layer: https://maskray.me/blog/2021-08-08-toolchain-testing#the-test-checks-at-the-wrong-layer

This should just test that clang passes -target-feature. It's unnecessary to enumerate every combination.

Also, use `--target=` for new tests. `-target ` has been deprecated since 3.4.

nickdesaulniers replied:
Yeah, a full on integration test for this is a bit much.  The backend tests LGTM; just making sure the frontend doesn't error when presented with this flag (and the `-no` variant) would be sufficient IMO.

nickdesaulniers commented:
Thanks for the patch!

stefanp-synopsys committed:
Changed the patch to use CodeGenOpts instead of target features.

I simplified the clang test significantly.
I also completely removed the llc test. Since the code gen opt is now internal
there is no good way that I know of to test it using llc.

stefanp-synopsys committed:
Fixed the broken test case.

stefanp-synopsys committed:
Fixed the formatting.

nickdesaulniers reviewed (approved):
LGTM; might be nice if it's possible to set this for `opt`/`llc` invocations, but for now I'm just happy to have this in clang.

MaskRay commented on clang/include/clang/Basic/CodeGenOptions.def:
```diff
@@ -197,6 +197,7 @@ CODEGENOPT(HIPCorrectlyRoundedDivSqrt, 1, 1) ///< -fno-hip-fp32-correctly-rounde
 CODEGENOPT(HIPSaveKernelArgName, 1, 0) ///< Set when -fhip-kernel-arg-name is enabled.
 CODEGENOPT(UniqueInternalLinkageNames, 1, 0) ///< Internal Linkage symbols get unique names.
 CODEGENOPT(SplitMachineFunctions, 1, 0) ///< Split machine functions using profile information.
+CODEGENOPT(UseFullRegisterNames, 1, 0) ///< Print full register names in assembly -mregnames
```
`PPCUseFullRegisterNames` or `PPCFullRegisterNames`

This is rs6000-specific in gcc.

MaskRay commented on clang/lib/Driver/ToolChains/Clang.cpp:
```diff
@@ -5011,6 +5011,12 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
   }
 
+  if (const Arg *A =
+          Args.getLastArg(options::OPT_mregnames, options::OPT_mno_regnames)) {
```
This will make other targets accept the ppc-specific -mregnames

stefanp-synopsys committed:
Added PPC to variable name and added isPPC() guard.

stefanp-synopsys committed:
Have the mregnames option also add the percent char as that is what GCC does as well.

MaskRay commented on clang/test/CodeGen/PowerPC/ppc-full-reg-names.c:
```diff
@@ -0,0 +1,18 @@
+// REQUIRES: powerpc-registered-target
+// RUN: %clang -### -target powerpc-ibm-aix-xcoff -mcpu=pwr8 -O3 -mregnames \
+// RUN:   %s 2>&1 >/dev/null | FileCheck %s --check-prefix=FULLNAMES
+// RUN: %clang -### -target powerpc64-ibm-aix-xcoff -mcpu=pwr8 -O3 -mregnames \
```
> This should just test that clang passes -target-feature. It's unnecessary to enumerate every combination.

I think most RUN lines can be removed. `-O3` should be removed as well.

MaskRay reviewed (approved):
Mostly looks good

chenzheng1030 commented on clang/include/clang/Basic/CodeGenOptions.def:
```diff
@@ -197,6 +197,7 @@ CODEGENOPT(HIPCorrectlyRoundedDivSqrt, 1, 1) ///< -fno-hip-fp32-correctly-rounde
 CODEGENOPT(HIPSaveKernelArgName, 1, 0) ///< Set when -fhip-kernel-arg-name is enabled.
 CODEGENOPT(UniqueInternalLinkageNames, 1, 0) ///< Internal Linkage symbols get unique names.
 CODEGENOPT(SplitMachineFunctions, 1, 0) ///< Split machine functions using profile information.
+CODEGENOPT(PPCUseFullRegisterNames, 1, 0) ///< Print full register names in assembly -mregnames
```
Nit: the last `-mregnames` seems unnecessary?

chenzheng1030 reviewed (approved):
Thanks for adding this. LGTM

stefanp-synopsys committed:
Address final comments on patch.

brad0 commented:
Thanks.
